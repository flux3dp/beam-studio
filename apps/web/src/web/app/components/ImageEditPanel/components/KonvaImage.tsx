/* eslint-disable no-underscore-dangle */
import React, { forwardRef, useEffect, useImperativeHandle, useRef } from 'react';

import Konva from 'konva';
import { Filter } from 'konva/lib/Node';
import { Image } from 'react-konva';
import useImage from 'use-image';

interface Props {
  src: string;
  filters?: Array<Filter>;
}

export type KonvaImageRef = Konva.Image & { useImageStatus: 'loading' | 'loaded' | 'failed' };

const KonvaImage = forwardRef<KonvaImageRef, Props>(({ src, filters }, ref) => {
  const [image, useImageStatus] = useImage(src, 'anonymous');
  const imageRef = useRef<Konva.Image>(null);

  useImperativeHandle(
    ref,
    () =>
      Object.assign(imageRef.current, {
        useImageStatus,
        // these methods are under 3 levels of prototype chain, so expose them explicitly
        isCached: imageRef.current.isCached,
        _getCachedSceneCanvas: imageRef.current._getCachedSceneCanvas,
      }),
    [useImageStatus]
  );

  useEffect(() => {
    if (image) {
      imageRef.current.cache({ pixelRatio: 1 });
    }
  }, [image]);

  return <Image ref={imageRef} image={image} filters={filters} />;
});

export default KonvaImage;
