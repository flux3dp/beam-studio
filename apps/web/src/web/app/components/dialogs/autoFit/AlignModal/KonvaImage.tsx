import React, { forwardRef, MutableRefObject, useEffect, useRef, useState } from 'react';
import Konva from 'konva';
import useImage from 'use-image';
import { Image, Transformer } from 'react-konva';
import { setRotationAngle } from 'app/svgedit/transform/rotation';

import findDefs from 'app/svgedit/utils/findDef';
import NS from 'app/constants/namespaces';
import svgStringToCanvas from 'helpers/image/svgStringToCanvas';
import symbolMaker from 'helpers/symbol-maker';
import updateElementColor from 'helpers/color/updateElementColor';

import { ImageDimension } from './dimension';

const getImageUrl = async (
  element: SVGElement,
  bbox: { x: number; y: number; width: number; height: number }
): Promise<string> => {
  const svgDefs = findDefs();
  const { x, y, width, height } = bbox;
  const svgContent = document.getElementById('svgcontent') as unknown as SVGSVGElement;
  const clonedSvgContent = svgContent.cloneNode(true) as SVGSVGElement;
  clonedSvgContent.setAttribute('width', (width + 2).toString());
  clonedSvgContent.setAttribute('height', (height + 2).toString());
  clonedSvgContent.setAttribute('viewBox', `${x - 1} ${y - 1} ${width + 2} ${height + 2}`);
  const elementId = element.getAttribute('id');
  const clonedElement = clonedSvgContent.getElementById(elementId);
  setRotationAngle(clonedElement as SVGElement, 0);
  // keep title, filter and selected element, remove other elements
  const elementsToDelete = clonedSvgContent.querySelectorAll(
    `g.layer > *:not(filter):not(title):not(#${elementId})`
  );
  elementsToDelete.forEach((e) => e.remove());
  const defs = document.createElementNS(NS.SVG, 'defs');
  clonedSvgContent.appendChild(defs);
  const useElements =
    clonedElement.tagName === 'use'
      ? [clonedElement as SVGUseElement]
      : [...clonedElement.querySelectorAll('use')];

  useElements.forEach((useElement) => {
    symbolMaker.switchImageSymbol(useElement, false);
    const href = useElement.getAttribute('href') || useElement.getAttribute('xlink:href');
    const symbol = svgDefs.querySelector(href);
    if (!symbol) return;
    const clonedSymbol = symbol.cloneNode(true) as SVGSymbolElement;
    defs.appendChild(clonedSymbol);
    updateElementColor(useElement);
  });
  const svgStr = new XMLSerializer().serializeToString(clonedSvgContent);
  const canvas = await svgStringToCanvas(svgStr, width + 2, height + 2);
  return canvas.toDataURL();
};

interface Props {
  isDragging: boolean;
  element: SVGElement;
  elementBBox: { x: number; y: number; width: number; height: number };
  initDimension: ImageDimension;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  onChange: (imageDimension: Partial<ImageDimension>) => void;
}

const KonvaImage = forwardRef<Konva.Image, Props>(
  (
    {
      isDragging,
      element,
      elementBBox,
      initDimension,
      onChange,
      onMouseEnter,
      onMouseLeave,
    }: Props,
    imageRef: MutableRefObject<Konva.Image>
  ): JSX.Element => {
    const [imageUrl, setImageUrl] = useState<string | null>(null);
    const [image, status] = useImage(imageUrl);
    const transformerInited = useRef(false);
    const transformerRef = useRef<Konva.Transformer>(null);

    useEffect(() => {
      getImageUrl(element, elementBBox).then(setImageUrl);
    }, [element, elementBBox]);

    useEffect(() => {
      if (!transformerInited.current && imageRef.current && transformerRef.current) {
        transformerRef.current.nodes([imageRef.current]);
        transformerInited.current = true;
      }
    });

    const { x, y, width, height, rotation } = initDimension;
    if (!image || status !== 'loaded') return null;

    return (
      <>
        <Image
          ref={imageRef}
          draggable={!isDragging}
          image={image}
          x={x}
          y={y}
          width={width}
          height={height}
          rotation={rotation}
          cursor="move"
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          onDragMove={(e) => {
            onChange({ x: e.target.x(), y: e.target.y() });
          }}
          onDragEnd={(e) => {
            onChange({ x: e.target.x(), y: e.target.y() });
          }}
          onTransform={() => {
            const node = imageRef.current;
            const scaleX = node.scaleX();
            const scaleY = node.scaleY();
            onChange({
              x: node.x(),
              y: node.y(),
              width: node.width() * scaleX,
              height: node.height() * scaleY,
              rotation: node.rotation(),
            });
          }}
          onTransformEnd={() => {
            const node = imageRef.current;
            const scaleX = node.scaleX();
            const scaleY = node.scaleY();

            node.scaleX(1);
            node.scaleY(1);
            node.width(node.width() * scaleX);
            node.height(node.height() * scaleY);
            onChange({
              x: node.x(),
              y: node.y(),
              width: node.width(),
              height: node.height(),
              rotation: node.rotation(),
            });
          }}
        />
        <Transformer
          ref={transformerRef}
          flipEnabled={false}
          anchorSize={8}
          anchorStyleFunc={(anchor) => {
            anchor.width(8);
            anchor.height(8);
            anchor.strokeWidth(2);
            anchor.cornerRadius(4);
            if (anchor.hasName('rotater')) {
              anchor.fill('#12b700');
              anchor.stroke('#0000ff');
            } else {
              anchor.stroke('#000000');
            }
          }}
          borderStroke="#0000ff"
          borderDash={[5, 5]}
          rotateAnchorCursor="url(js/lib/svgeditor/images/rotate.png) 12 12, auto"
          rotateAnchorOffset={20}
        />
      </>
    );
  }
);

export default KonvaImage;
