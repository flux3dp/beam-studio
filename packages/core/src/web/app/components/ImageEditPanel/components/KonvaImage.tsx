import React, { forwardRef, useEffect, useImperativeHandle, useRef } from 'react';

import type Konva from 'konva';
import type { Filter } from 'konva/lib/Node';
import { Image } from 'react-konva';
import useImage from 'use-image';

interface Props {
  cornerRadius?: number;
  filters?: Filter[];
  src: string;
}

export type KonvaImageRef = Konva.Image & { useImageStatus: 'failed' | 'loaded' | 'loading' };

const KonvaImage = forwardRef<KonvaImageRef, Props>(({ cornerRadius = 0, filters, src }, ref) => {
  const [image, useImageStatus] = useImage(src, 'anonymous');
  const imageRef = useRef<Konva.Image>(null);

  useImperativeHandle(
    ref,
    () =>
      Object.assign(imageRef.current!, {
        _getCachedSceneCanvas: imageRef.current!._getCachedSceneCanvas,
        // these methods are under 3 levels of prototype chain, so expose them explicitly
        isCached: imageRef.current!.isCached,
        useImageStatus,
      }),
    [useImageStatus],
  );

  useEffect(() => {
    if (image) {
      imageRef.current!.cache({ pixelRatio: 1 });
    }
    // force redraw when image or cornerRadius changes
  }, [image, cornerRadius]);

  return <Image cornerRadius={cornerRadius} filters={filters} image={image} ref={imageRef} />;
});

export default KonvaImage;
