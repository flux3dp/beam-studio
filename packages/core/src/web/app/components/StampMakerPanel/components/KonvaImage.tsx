import React, { forwardRef, useEffect, useImperativeHandle, useRef } from 'react';

import type Konva from 'konva';
import type { Filter } from 'konva/lib/Node';
import { Image } from 'react-konva';
import useImage from 'use-image';

interface Props {
  filters?: Filter[];
  horizontalFlip?: boolean;
  src: string;
}

export type KonvaImageRef = Konva.Image & { useImageStatus: 'failed' | 'loaded' | 'loading' };

const KonvaImage = forwardRef<KonvaImageRef, Props>(({ filters, horizontalFlip = false, src }, ref) => {
  const [image, useImageStatus] = useImage(src, 'anonymous');
  const imageRef = useRef<Konva.Image>(null);

  useImperativeHandle(
    ref,
    () =>
      Object.assign(imageRef.current!, {
        _getCachedSceneCanvas: imageRef.current!._getCachedSceneCanvas,
        // these methods are under 3 levels of prototype chain, so expose them explicitly
        isCached: imageRef.current!.isCached,
        useImageStatus,
      }),
    [useImageStatus],
  );

  useEffect(() => {
    if (image && imageRef.current) {
      // Apply horizontal flip transformation
      const imageNode = imageRef.current;

      imageNode.scaleX(horizontalFlip ? -1 : 1);
      imageNode.offsetX(horizontalFlip ? image.width : 0);
      imageNode.cache({ pixelRatio: 1 });
    }
    // force redraw when image or transformations change
  }, [image, horizontalFlip]);

  // <Text align=''></Text>
  // <Image f />;

  return <Image fill={'white'} filters={filters} image={image} ref={imageRef} />;
});

export default KonvaImage;
