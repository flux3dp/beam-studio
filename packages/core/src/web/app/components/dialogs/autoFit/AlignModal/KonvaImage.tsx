import type { MutableRefObject } from 'react';
import React, { forwardRef, useEffect, useRef, useState } from 'react';

import type Konva from 'konva';
import { Image, Transformer } from 'react-konva';
import useImage from 'use-image';

import NS from '@core/app/constants/namespaces';
import { setRotationAngle } from '@core/app/svgedit/transform/rotation';
import findDefs from '@core/app/svgedit/utils/findDef';
import updateElementColor from '@core/helpers/color/updateElementColor';
import svgStringToCanvas from '@core/helpers/image/svgStringToCanvas';
import symbolMaker from '@core/helpers/symbol-maker';

import type { ImageDimension } from './dimension';

const getImageUrl = async (
  element: SVGElement,
  bbox: { height: number; width: number; x: number; y: number },
): Promise<string> => {
  const svgDefs = findDefs();
  const { height, width, x, y } = bbox;
  const svgContent = document.getElementById('svgcontent') as unknown as SVGSVGElement;
  const clonedSvgContent = svgContent.cloneNode(true) as SVGSVGElement;

  clonedSvgContent.setAttribute('width', (width + 2).toString());
  clonedSvgContent.setAttribute('height', (height + 2).toString());
  clonedSvgContent.setAttribute('viewBox', `${x - 1} ${y - 1} ${width + 2} ${height + 2}`);

  const elementId = element.getAttribute('id');
  const clonedElement = clonedSvgContent.getElementById(elementId);

  setRotationAngle(clonedElement as SVGElement, 0);

  // keep title, filter and selected element, remove other elements
  const elementsToDelete = clonedSvgContent.querySelectorAll(`g.layer > *:not(filter):not(title):not(#${elementId})`);

  elementsToDelete.forEach((e) => e.remove());

  const defs = document.createElementNS(NS.SVG, 'defs');

  clonedSvgContent.appendChild(defs);

  const useElements =
    clonedElement.tagName === 'use' ? [clonedElement as SVGUseElement] : [...clonedElement.querySelectorAll('use')];

  useElements.forEach((useElement) => {
    symbolMaker.switchImageSymbol(useElement, false);

    const href = useElement.getAttribute('href') || useElement.getAttribute('xlink:href');
    const symbol = svgDefs.querySelector(href);

    if (!symbol) {
      return;
    }

    const clonedSymbol = symbol.cloneNode(true) as SVGSymbolElement;

    defs.appendChild(clonedSymbol);
    updateElementColor(useElement);
  });

  const svgStr = new XMLSerializer().serializeToString(clonedSvgContent);
  const canvas = await svgStringToCanvas(svgStr, width + 2, height + 2);

  return canvas.toDataURL();
};

interface Props {
  element: SVGElement;
  elementBBox: { height: number; width: number; x: number; y: number };
  initDimension: ImageDimension;
  isDragging: boolean;
  onChange: (imageDimension: Partial<ImageDimension>) => void;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
}

const KonvaImage = forwardRef<Konva.Image, Props>(
  (
    { element, elementBBox, initDimension, isDragging, onChange, onMouseEnter, onMouseLeave }: Props,
    imageRef: MutableRefObject<Konva.Image>,
  ): React.JSX.Element => {
    const [imageUrl, setImageUrl] = useState<null | string>(null);
    const [image, status] = useImage(imageUrl);
    const transformerInited = useRef(false);
    const transformerRef = useRef<Konva.Transformer>(null);

    useEffect(() => {
      getImageUrl(element, elementBBox).then(setImageUrl);
    }, [element, elementBBox]);

    useEffect(() => {
      if (!transformerInited.current && imageRef.current && transformerRef.current) {
        transformerRef.current.nodes([imageRef.current]);
        transformerInited.current = true;
      }
    });

    const { height, rotation, width, x, y } = initDimension;

    if (!image || status !== 'loaded') {
      return null;
    }

    return (
      <>
        <Image
          cursor="move"
          draggable={!isDragging}
          height={height}
          image={image}
          onDragEnd={(e) => {
            onChange({ x: e.target.x(), y: e.target.y() });
          }}
          onDragMove={(e) => {
            onChange({ x: e.target.x(), y: e.target.y() });
          }}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          onTransform={() => {
            const node = imageRef.current;
            const scaleX = node.scaleX();
            const scaleY = node.scaleY();

            onChange({
              height: node.height() * scaleY,
              rotation: node.rotation(),
              width: node.width() * scaleX,
              x: node.x(),
              y: node.y(),
            });
          }}
          onTransformEnd={() => {
            const node = imageRef.current;
            const scaleX = node.scaleX();
            const scaleY = node.scaleY();

            node.scaleX(1);
            node.scaleY(1);
            node.width(node.width() * scaleX);
            node.height(node.height() * scaleY);
            onChange({
              height: node.height(),
              rotation: node.rotation(),
              width: node.width(),
              x: node.x(),
              y: node.y(),
            });
          }}
          ref={imageRef}
          rotation={rotation}
          width={width}
          x={x}
          y={y}
        />
        <Transformer
          anchorSize={8}
          anchorStyleFunc={(anchor) => {
            anchor.width(8);
            anchor.height(8);
            anchor.strokeWidth(2);
            anchor.cornerRadius(4);

            if (anchor.hasName('rotater')) {
              anchor.fill('#12b700');
              anchor.stroke('#0000ff');
            } else {
              anchor.stroke('#000000');
            }
          }}
          borderDash={[5, 5]}
          borderStroke="#0000ff"
          flipEnabled={false}
          ref={transformerRef}
          rotateAnchorCursor="url(js/lib/svgeditor/images/rotate.png) 12 12, auto"
          rotateAnchorOffset={20}
        />
      </>
    );
  },
);

export default KonvaImage;
