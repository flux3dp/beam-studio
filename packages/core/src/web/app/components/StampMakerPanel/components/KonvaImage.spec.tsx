import React from 'react';
import { render } from '@testing-library/react';

const mockKonvaFilters = {
  Blur: jest.fn(),
  Brighten: jest.fn(),
  Invert: jest.fn(),
};

jest.mock('konva', () => ({
  Filters: mockKonvaFilters,
}));

const mockUseImage = jest.fn();
const mockCache = jest.fn();
const mockScaleX = jest.fn();
const mockOffsetX = jest.fn();
const mockIsCached = jest.fn();
const mockGetCachedSceneCanvas = jest.fn();

jest.mock('use-image', () => mockUseImage);

jest.mock('react-konva', () => ({
  Image: React.forwardRef(({ fill, filters, image }: any, ref: any) => {
    React.useImperativeHandle(ref, () => ({
      _getCachedSceneCanvas: mockGetCachedSceneCanvas,
      cache: mockCache,
      isCached: mockIsCached,
      offsetX: mockOffsetX,
      scaleX: mockScaleX,
    }));

    return (
      <div data-testid="konva-image">
        <span>Filters: {filters?.length || 0}</span>
        <span>Image: {image ? 'loaded' : 'not loaded'}</span>
        <span>Fill: {fill}</span>
      </div>
    );
  }),
}));

import type { KonvaImageRef } from './KonvaImage';
import KonvaImage from './KonvaImage';

describe('test KonvaImage', () => {
  const mockImage = {
    height: 100,
    width: 100,
  } as HTMLImageElement;

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseImage.mockReturnValue([mockImage, 'loaded']);
    mockIsCached.mockReturnValue(true);
  });

  it('should render correctly', () => {
    const { container } = render(<KonvaImage src="test.png" />);

    expect(container).toMatchSnapshot();
  });

  it('should load image with useImage hook', () => {
    render(<KonvaImage src="test.png" />);

    expect(mockUseImage).toHaveBeenCalledWith('test.png', 'anonymous');
  });

  it('should render with filters prop', () => {
    const filters = [mockKonvaFilters.Blur, mockKonvaFilters.Brighten];
    const { getByText } = render(<KonvaImage filters={filters} src="test.png" />);

    expect(getByText('Filters: 2')).toBeInTheDocument();
  });

  it('should apply horizontal flip when horizontalFlip is true', () => {
    render(<KonvaImage horizontalFlip={true} src="test.png" />);

    expect(mockScaleX).toHaveBeenCalledWith(-1);
    expect(mockOffsetX).toHaveBeenCalledWith(100);
    expect(mockCache).toHaveBeenCalledWith({ pixelRatio: 1 });
  });

  it('should not apply horizontal flip when horizontalFlip is false', () => {
    render(<KonvaImage horizontalFlip={false} src="test.png" />);

    expect(mockScaleX).toHaveBeenCalledWith(1);
    expect(mockOffsetX).toHaveBeenCalledWith(0);
    expect(mockCache).toHaveBeenCalledWith({ pixelRatio: 1 });
  });

  it('should use default horizontalFlip value of false', () => {
    render(<KonvaImage src="test.png" />);

    expect(mockScaleX).toHaveBeenCalledWith(1);
    expect(mockOffsetX).toHaveBeenCalledWith(0);
  });

  it('should expose ref methods correctly', () => {
    const ref = React.createRef<KonvaImageRef>();

    render(<KonvaImage ref={ref} src="test.png" />);

    expect(ref.current).toBeDefined();
    expect(ref.current?.isCached).toBe(mockIsCached);
    expect(ref.current?._getCachedSceneCanvas).toBe(mockGetCachedSceneCanvas);
    expect(ref.current?.useImageStatus).toBe('loaded');
  });

  it('should update when image loading status changes', () => {
    mockUseImage.mockReturnValue([null, 'loading']);

    const { getByText, rerender } = render(<KonvaImage src="test.png" />);

    expect(getByText('Image: not loaded')).toBeInTheDocument();

    mockUseImage.mockReturnValue([mockImage, 'loaded']);
    rerender(<KonvaImage src="test.png" />);

    expect(getByText('Image: loaded')).toBeInTheDocument();
  });

  it('should not cache when image is not loaded', () => {
    mockUseImage.mockReturnValue([null, 'loading']);
    render(<KonvaImage src="test.png" />);

    expect(mockCache).not.toHaveBeenCalled();
  });

  it('should re-cache when horizontalFlip changes', () => {
    const { rerender } = render(<KonvaImage horizontalFlip={false} src="test.png" />);

    expect(mockCache).toHaveBeenCalledTimes(1);

    rerender(<KonvaImage horizontalFlip={true} src="test.png" />);

    expect(mockCache).toHaveBeenCalledTimes(2);
    expect(mockScaleX).toHaveBeenLastCalledWith(-1);
    expect(mockOffsetX).toHaveBeenLastCalledWith(100);
  });

  it('should handle image loading failure', () => {
    mockUseImage.mockReturnValue([null, 'failed']);

    const ref = React.createRef<KonvaImageRef>();

    render(<KonvaImage ref={ref} src="test.png" />);

    expect(ref.current?.useImageStatus).toBe('failed');
    expect(mockCache).not.toHaveBeenCalled();
  });

  it('should set fill to white', () => {
    const { getByText } = render(<KonvaImage src="test.png" />);

    expect(getByText('Fill: white')).toBeInTheDocument();
  });
});
