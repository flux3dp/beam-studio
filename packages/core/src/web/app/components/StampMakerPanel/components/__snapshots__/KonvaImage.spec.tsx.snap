// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test KonvaImage should render correctly 1`] = `
<div>
  <div
    data-testid="konva-image"
  >
    <span>
      Filters: 
      0
    </span>
    <span>
      Image: 
      loaded
    </span>
    <span>
      Fill: 
      white
    </span>
  </div>
</div>
`;
